clippy_lints/src/utils/hir_utils.rs

```diff
@@ -673,4 +673,100 @@ impl<'a, 'tcx: 'a> SpanlessHash<'a, 'tcx> {
+            TyKind::Slice(ref ty) => {
+                self.hash_ty(ty);
+            },
+            TyKind::Array(ref ty, anon_const) => {
+                self.hash_ty(ty);
+                self.hash_expr(&self.cx.tcx.hir().body(anon_const.body).value);
+            },
+            TyKind::Ptr(ref mut_ty) => {
+                self.hash_ty(&mut_ty.ty);
+                mut_ty.mutbl.hash(&mut self.s);
+            },
+            TyKind::Rptr(ref lifetime, ref mut_ty) => {
+                self.hash_lifetime(lifetime);
+                self.hash_ty(&mut_ty.ty);
+                mut_ty.mutbl.hash(&mut self.s);
+            },
+            TyKind::BareFn(ref bfn) => {
+                bfn.unsafety.hash(&mut self.s);
+                bfn.abi.hash(&mut self.s);
+                for arg in &bfn.decl.inputs {
+                    self.hash_ty(&arg);
+                }
+                match bfn.decl.output {
+                    FunctionRetTy::DefaultReturn(_) => {
+                        ().hash(&mut self.s);
+                    },
+                    FunctionRetTy::Return(ref ty) => {
+                        self.hash_ty(ty);
+                    },
+                }
+                bfn.decl.c_variadic.hash(&mut self.s);
+            },
+            TyKind::Tup(ref ty_list) => {
+                for ty in ty_list {
+                    self.hash_ty(ty);
+                }
+
+            },
+            TyKind::Path(ref qpath) => {
+                match qpath {
+                    QPath::Resolved(ref maybe_ty, ref path) => {
+                        if let Some(ref ty) = maybe_ty {
+                            self.hash_ty(ty);
+                        }
+                        for segment in &path.segments {
+                            segment.ident.name.hash(&mut self.s);
+                        }
+                    }, 
+                    QPath::TypeRelative(ref ty, ref segment) => {
+                        self.hash_ty(ty);
+                        segment.ident.name.hash(&mut self.s);
+                    },
+                }
+            },
+            TyKind::Def(_, ref arg_list) => {
+                for arg in arg_list {
+                    match arg {
+                        GenericArg::Lifetime(ref l) => self.hash_lifetime(l),
+                        GenericArg::Type(ref ty) => self.hash_ty(ty),
+                        GenericArg::Const(ref ca) => {
+                            self.hash_expr(&self.cx.tcx.hir().body(ca.value.body).value);
+                        },
+                    }
+                }
+            },
+            TyKind::TraitObject(_, ref lifetime) => {
+                self.hash_lifetime(lifetime);
+
+            },
+            TyKind::Typeof(ref anon_const) => {
+                self.hash_expr(&self.cx.tcx.hir().body(anon_const.body).value);
+            },
+            TyKind::CVarArgs(ref lifetime) => {
+                self.hash_lifetime(lifetime);
+            },
+            _ => {},
```

Comment:
What `TyKind`s are missing here? I think it would be better to make this exhaustive, instead of using the wildcard pattern.

Reply:
`Err`, `Infer` and `Never`, so just the ones with no other data contained. I'll change the match so that's clear